### PR TITLE
Threading fixes

### PR DESCRIPTION
**Pull request by DenizPiri**
## Description

This PR fixes many minor issues with threading.
Also fixes an issue with RoundRobinRouterState where it would fail beyond int.MaxValue messages.

## Purpose
This pull request is a:

- [X] Bugfix (non-breaking change which fixes an issue)

## Checklist

- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by rogeralsing** [comment]
Many thanks @DenizPiri ! this is great!

File: src/Proto.Actor/Future/SharedFuture.cs [hunk 1]
```diff
@@ -70,7 +70,15 @@ internal SharedFutureProcess(ActorSystem system, int size) : base(system)
         private PID Pid { get; }
         public bool Stopping { get; private set; }
 
-        public int RequestsInFlight => (int) (_createdRequests - _completedRequests);
+        public int RequestsInFlight {
+            get {
+                // Read completedRequests first and createdRequests later so that we will
+                // never read the 2 vars in an order that would result in completedRequests > createdRequests.
+                long completed = Interlocked.Read(ref _createdRequests);
```
Comment (swimtver) [inline src/Proto.Actor/Future/SharedFuture.cs:77]: looks like here should be _completedRequests.

Comment (rogeralsing) [reply src/Proto.Actor/Future/SharedFuture.cs:77]: Yes, you are correct. thanks for spotting it

Comment (DenizPiri) [reply src/Proto.Actor/Future/SharedFuture.cs:77]: Seems like this is already handled by mhelleborg. Resolving this.

**Pull request merged**